app/src/main/cpp/hash.c

```diff
@@ -0,0 +1,57 @@
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <android/log.h>
+#include <jni.h>
+#include <sha1.h>
+
+#include "ehviewer.h"
+
+#define LOG_TAG "hash"
+
+#define BUFFER_SIZE 8192
+
+typedef uint8_t byte;
+
+const char hex_digits[] = "0123456789abcdef";
+
+char *bytes_to_hex(const byte *bytes, size_t length) {
+    byte byte;
+    char *hex = malloc(2 * length + 1);
+    for (int i = 0; i < length; i++) {
+        byte = bytes[i];
+        hex[2 * i] = hex_digits[byte >> 4 & 0xF];
+        hex[2 * i + 1] = hex_digits[byte & 0xF];
+    }
+    hex[2 * length] = '\0';
+    return hex;
+}
+
+JNIEXPORT jstring JNICALL
+Java_com_hippo_ehviewer_jni_HashKt_sha1(JNIEnv *env, jclass clazz, jint fd) {
+    EH_UNUSED(clazz);
+    byte *buffer = malloc(BUFFER_SIZE);
```

Comment:
Allocate buffer on stack

Reply:
Should inline `bytes_to_hex` as well I guess, it most likely won't be used anywhere else.